blas/gonum/dgemm.go

```diff
@@ -21,25 +21,75 @@ import (
 // an m×n matrix, and alpha and beta are scalars. tA and tB specify whether A or
 // B are transposed.
 func (Implementation) Dgemm(tA, tB blas.Transpose, m, n, k int, alpha float64, a []float64, lda int, b []float64, ldb int, beta float64, c []float64, ldc int) {
-	if tA != blas.NoTrans && tA != blas.Trans && tA != blas.ConjTrans {
+	switch tA {
+	default:
 		panic(badTranspose)
+	case blas.NoTrans, blas.Trans, blas.ConjTrans:
 	}
-	if tB != blas.NoTrans && tB != blas.Trans && tB != blas.ConjTrans {
+	switch tB {
+	default:
 		panic(badTranspose)
+	case blas.NoTrans, blas.Trans, blas.ConjTrans:
+	}
+	if m < 0 {
+		panic(mLT0)
+	}
+	if n < 0 {
+		panic(nLT0)
+	}
+	if k < 0 {
+		panic(kLT0)
 	}
 	aTrans := tA == blas.Trans || tA == blas.ConjTrans
 	if aTrans {
-		checkDMatrix('a', k, m, a, lda)
+		if lda < max(1, m) {
+			panic(badLdA)
+		}
 	} else {
-		checkDMatrix('a', m, k, a, lda)
+		if lda < max(1, k) {
+			panic(badLdA)
+		}
 	}
 	bTrans := tB == blas.Trans || tB == blas.ConjTrans
 	if bTrans {
-		checkDMatrix('b', n, k, b, ldb)
+		if ldb < max(1, k) {
+			panic(badLdB)
+		}
 	} else {
-		checkDMatrix('b', k, n, b, ldb)
+		if ldb < max(1, n) {
+			panic(badLdB)
+		}
+	}
+	if ldc < max(1, n) {
+		panic(badLdC)
+	}
+
+	// Quick return if possible.
+	if m == 0 || n == 0 || ((alpha == 0 || k == 0) && beta == 1) {
+		return
+	}
+
+	if aTrans {
```

Comment:
Can you add a comment to the cases where you have moved checks to after quick returns explaining that the failure is not possible for the quick return case (with nicer words than I have access to right now obviously).

Also for other commits where these checks have moved.

Reply:
As if I knew how to say it with nicer words :-)

But looking at the code, I think I will change some of the quick return checks before merging. Using `Zgbmv` as an example, I'll change it to:
```
if m == 0 || n == 0 {
    return
}
// Length checks
if alpha == 0 && beta == 1 {
    return
}

Reply:
How about
```
// For zero matrix size the following slice length checks are trivially satisfied.
```
?

Reply:
SGTM

Reply:
>As if I knew how to say it with nicer words :-)

You do the best documentation.